### PR TITLE
Fix for connection problemens / comm errors

### PR DESCRIPTION
**Pull request by vanelsberg**
In essence this fix (more like a work around) solves unexpected behavior of the BL library where reconnecting to RL results in a connective state where communications are impossible. 

As a result OmniCore kept reconnecting (EnsureDevice) en trying to send messages to RL with no avail.
Only way to get out of this was forcefully restarting OmniCore and BT toggling.

The fix disposes of and recreate the RileyLink instance as soon as it gets in this unexpected state to reinitializing RL and restore communications.

### CODE REVIEW

File: OmniCore.Mobile/OmniCore.Mobile.Android/MainActivity.cs [hunk 1]
```diff
@@ -53,6 +55,27 @@ protected override void OnCreate(Bundle savedInstanceState)
...
+        {
+            try
+            {
+                if (CrossBleAdapter.Current.CanControlAdapterState())
+                {
+                    // Switch adapter on if off (testing)
+                    if (CrossBleAdapter.Current.Status == AdapterStatus.PoweredOff)
+                    {
+                        CrossBleAdapter.Current.SetAdapterState(true);
+                        Wait(5000);
```
Comment (winemug) [inline OmniCore.Mobile/OmniCore.Mobile.Android/MainActivity.cs:68]: why are we waiting here? and what if the user declines the request - further checks? shutdown app? etc.

File: OmniCore.Mobile/OmniCore.Mobile.Android/Properties/AndroidManifest.xml [hunk 2]
```diff
@@ -3,4 +3,6 @@
 	<uses-sdk android:minSdkVersion="23" android:targetSdkVersion="28" />
 	<application android:label="OmniCore" android:largeHeap="true"></application>
 	<uses-feature android:name="android.hardware.bluetooth_le" android:required="false" />
+	<uses-permission android:name="android.permission.BLUETOOTH"/>
```
Comment (winemug) [inline OmniCore.Mobile/OmniCore.Mobile.Android/Properties/AndroidManifest.xml:6]: these are not necessary, xamarin framework compiles the attributes in the assembly info into the manifest

File: OmniCore.Model/OmniCore.Model/Interfaces/IMessageExchangeProvider.cs [hunk 3]
```diff
@@ -10,6 +10,6 @@ namespace OmniCore.Model.Interfaces
 {
     public interface IMessageExchangeProvider
     {
-        Task<IMessageExchange> GetMessageExchange(IMessageExchangeParameters messageExchangeParameters, IPod pod);
+        Task<IMessageExchange> GetMessageExchange(IMessageExchangeParameters messageExchangeParameters, IPod pod, IMessageExchangeProgress messageProgress);
```
Comment (winemug) [inline OmniCore.Model/OmniCore.Model/Interfaces/IMessageExchangeProvider.cs:13]: design here changed already significantly, is incompatible with current code

File: OmniCore.Model/OmniCore.Model.Eros/ErosPodManager.cs [hunk 4]
```diff
@@ -134,7 +134,7 @@ private ErosMessageExchangeParameters GetStandardParameters()
                 progress.ActionText = "Started new message exchange";
                 progress.Result.RequestTime = DateTimeOffset.UtcNow;
                 progress.Running = true;
-                var messageExchange = await MessageExchangeProvider.GetMessageExchange(messageExchangeParameters, Pod);
+                var messageExchange = await MessageExchangeProvider.GetMessageExchange(messageExchangeParameters, Pod, progress);
```
Comment (winemug) [inline OmniCore.Model/OmniCore.Model.Eros/ErosPodManager.cs:137]: see comment about design, this class is no more

File: OmniCore.Radio/OmniCore.Radio.RileyLink/RileyLinkMessageExchangeProvider.cs [hunk 5]
```diff
@@ -17,8 +17,23 @@ public class RileyLinkMessageExchangeProvider : IMessageExchangeProvider
         private static RileyLinkMessageExchange RileyLinkMessageExchange;
 
         [method: SuppressMessage("", "CS1998", Justification = "Not applicable")]
-        public async Task<IMessageExchange> GetMessageExchange(IMessageExchangeParameters messageExchangeParameters, IPod pod)
+        public async Task<IMessageExchange> GetMessageExchange(IMessageExchangeParameters messageExchangeParameters, IPod pod, IMessageExchangeProgress messageProgress)
```
Comment (winemug) [inline OmniCore.Radio/OmniCore.Radio.RileyLink/RileyLinkMessageExchangeProvider.cs:20]: design conflict again, messageexchange contains the progress not the other way around. I like the idea of provider verifying connectivity, but in that case more changes are needed - the way it is, it violates the separation of concerns principle

File: OmniCore.Radio/OmniCore.Radio.RileyLink/RileyLink.cs [hunk 6]
```diff
@@ -54,7 +54,7 @@ private RileyLinkStatistics GetStatistics(IMessageExchangeProgress messageProgre
 
         public RileyLink(ErosRadioPreferences erosRadioPreferences)
         {
-            TxAmplification = TxPower.A4_Normal;
+            TxAmplification = TxPower.A5_High; // On initialisation, start with some extra TxPower (it will auto adjust)
```
Comment (winemug) [inline OmniCore.Radio/OmniCore.Radio.RileyLink/RileyLink.cs:57]: I don't see how this is justified, high tx is one of the primary reasons for pod failures as it leads to broken comms

File: OmniCore.Radio/OmniCore.Radio.RileyLink/RileyLink.cs [hunk 7]
```diff
@@ -176,6 +176,50 @@ private async Task ConnectToDevice(IMessageExchangeProgress messageProgress)
             }
         }
 
+        /// <summary>
+        /// Disconnect/Cancel active connection
+        /// </summary>
+        /// <param name="messageProgress">Messageprogress object</param>
+        public void DisconnectDevice(IMessageExchangeProgress messageProgress)
```
Comment (winemug) [inline OmniCore.Radio/OmniCore.Radio.RileyLink/RileyLink.cs:183]: I don't see this method called anywhere

**Review by winemug — COMMENTED** [review]
I am confused as to what this is fixing, as there is practically no change to the behavior -- The EnsureDevice method does the same in the newly added method already (and more) so I either miss the point completely - or you started on a not-up-to-date version which was significantly different?

Edit: I just realized not all commits were pushed to the repo, so please have a look at the latest state. It's still work in progress though.

File: OmniCore.Mobile/OmniCore.Mobile.Android/MainActivity.cs (hunk 1 above)
Comment (vanelsberg) [reply OmniCore.Mobile/OmniCore.Mobile.Android/MainActivity.cs:68]: Yes. * Please ignore! *
Actually a bit of test code, looking for a way to "auto-toggle" BT as this seemed to way to regain connectivity.

Left it in as I wanted to leave the code identical to what i tested with.

Btw: I do not see any user request. as wondering if this should be programmed additionally.

File: OmniCore.Mobile/OmniCore.Mobile.Android/Properties/AndroidManifest.xml (hunk 2 above)
Comment (vanelsberg) [reply OmniCore.Mobile/OmniCore.Mobile.Android/Properties/AndroidManifest.xml:6]: Ok. * Please also ignore *
Did not know. Found this somewhere in the documentation.
(Was part of my "test" for "auto-toggle" BT)

File: OmniCore.Model/OmniCore.Model/Interfaces/IMessageExchangeProvider.cs (hunk 3 above)
Comment (vanelsberg) [reply OmniCore.Model/OmniCore.Model/Interfaces/IMessageExchangeProvider.cs:13]: Base my changes on early 702 code from Git.
Was the easiest way to implement the change for rebuilding the RileyLinkInstance in GetMessageExchange().

Check.
As this is not compatible with current code, also ignore waiting for a more final resoluition.
(Actually I do not expect you to "merge" anything directly into the current branch.
Mainly trying to report my findings and give you some feedback on possible issues or ways to fix)

File: OmniCore.Model/OmniCore.Model.Eros/ErosPodManager.cs (hunk 4 above)
Comment (vanelsberg) [reply OmniCore.Model/OmniCore.Model.Eros/ErosPodManager.cs:137]: See e
Was the easiest way to implement the change for rebuilding the RileyLinkInstance in GetMessageExchange().

See my comments earlier on the same

File: OmniCore.Radio/OmniCore.Radio.RileyLink/RileyLinkMessageExchangeProvider.cs (hunk 5 above)
Comment (vanelsberg) [reply OmniCore.Radio/OmniCore.Radio.RileyLink/RileyLinkMessageExchangeProvider.cs:20]: Agree.
Was the easiest way to implement the change for rebuilding the RileyLinkInstance in GetMessageExchange(). Did not want to rewrite/redesign code.
Just "proof of concept" on "rebuilding RileyLinkInstance in GetMessageExchange()."

File: OmniCore.Radio/OmniCore.Radio.RileyLink/RileyLink.cs (hunk 6 above)
Comment (vanelsberg) [reply OmniCore.Radio/OmniCore.Radio.RileyLink/RileyLink.cs:57]: Considered this as safe:
1. Autoadjust will take it back to to normal or lower after initial comms?
2. Effectively it is identical to what I do when receiving errors on Pod comminucations: bringen RL (much) closer to the Pod?

Also: Autoadjust starts lowering transmission level on Pod communication errors - which I find contra-intuitive. Maybe start at a bit higher level to make sure range problems have less effect on the  initialization fase?

File: OmniCore.Radio/OmniCore.Radio.RileyLink/RileyLink.cs (hunk 7 above)
Comment (vanelsberg) [reply OmniCore.Radio/OmniCore.Radio.RileyLink/RileyLink.cs:183]: Check.
* Please ignore *
Test code...

**Comment by vanelsberg** [comment]
Strange! 
* The actual fix/work arround I was testing is not in! *

Somehow the actual code that improves reconnecting and/or make it easier to regain connectivity is not in this pull request. It somehow got lost with the update "Undo minor refactoring (keep identical code for previous test)" which is about removing two lines of code.
Will look into that.

**Comment by vanelsberg** [comment]
> I am confused as to what this is fixing, as there is practically no change to the behavior -- The  EnsureDevice method does the same in the newly added method already (and more) so I either miss the point completely - or you started on a not-up-to-date version which was significantly different?

This Pull request was to give you my changes on .702 while testing behavior. But it appears to me that only resulted in a lot of (understandable) confusion at your end :-)

**What actually "fixes"** or at least seems to gets me much more stability and a better way to manually restore connections without the need of restarting the application **is all about GetMessageExchange()**

_In the original implementation the RileyLinkInstance instance is initialized for the lifetime of the App.
No way to renew other then restarting the App._

**Why think it improves**:
I observed that somehow EnsureDevice() is unable the restore communication. So OmniCore keeps getting errors, never getting out of this state.

See DeviceIsValid():
Also observed that sometimes device state is not connected, on other times it is but EnsureDevice can not connect. This happens for instance when walking away whit RL from the Phone. On return RL goes green (BT connected) but somehow communication is then failing on EnsureDevice().

So I added this additional check to GetMessageExchange(). If it fails: dispose of current RileyLinkInstance an recreate. 

(See commit for "Check connection state." ):
```
           if (RileyLinkInstance != null)
            {
                // Make sure this instance is still valid
                if (!await RileyLinkInstance.DeviceIsValid(messageProgress))
                {
                    // Dispose() and create new RileyLinkInstance
                    RileyLinkInstance = null;
                    // Dispose() and create new RileyLinkMessageExchange
                    RileyLinkMessageExchange = null;
                }
            }

```

**Comment by vanelsberg** [comment]
Looking forward to your progress on OmniCore!
Will give it a test it as soon as you are ready. May then look again when comms are still troublesome.

My suggestion for the above was meant for sharing my findings and maybe give a quick workaround to improve 702 on returning form your holidays. But I get you been working hard beyond 702 :-)
So please, regard this as FYI only.

May I suggest you think about ways for self-recovering on the connection?
In my experience lifetime object instances can go into a failure state that is not repairable. The application should be able to detect this and either repair or do an emergency "shutdown"?

**Comment by vanelsberg** [comment]
Propose to close this pull request?

**Comment by winemug** [comment]
@vanelsberg many thanks for your extensive and insightful feedback. I believe the connection fixes are already addressed in a very similar manner as you have proposed but as I said earlier didn't make it to an alpha release. As we're moving to beta I am hoping to create a more stable basis so contributions like yours can be merged with ease.

**Pull request closed without merging**